### PR TITLE
[Rust] Fail unification instead of returning an error when type casts fail

### LINKED ISSUE #788: [Rust] Unifying different types throws an error
`?= new Foo() = new Bar()`

with two PolarClasses Foo and Bar throws:

```
Error: TypeError(TypeError { got: Some("Bar"), expected: "Foo" })
```

Instead of just failing to unify.

This is also problematic when used with `nil`. e.g. `?= not new Foo() = nil`

```
"error in inline query: Type error: Expected Foo got Option\n"
```

### PR DESCRIPTION
**Pull request by AnIrishDuck**
Fixes #788

I'm not sure this is the desired behavior. I couldn't find any tests for equivalent behavior in the Java or Go bindings.

However, once I was able to create a reproducible test case, the source of the behavior described in the original bug report was clear. It should be relatively easy to tweak depending on the ideal results.

### CODE REVIEW

**Comment by AnIrishDuck** [comment]
I have read the CLA Document and I hereby sign the CLA

**Review by samscott89 — COMMENTED** [review]
Thanks for opening this! The test case looks great.

One thing that I'm not sure about the implementation: the equality check currently expects a method `Fn(&T, &T) -> ...`.  

I'm wondering whether it's more internally consistent if we keep the error on the wrong types _here_, and instead check types elsewhere. For example, the caller `Class::equals` on line 112 in class.rs 

Alternatively (and the better functionality), we could allow setting equality checks for types `&T, &U`. But we don't currently have a mechanism for adding multiple equality checks.

I think we don't need to worry about the second point yet. This PR addresses the main problem raised in the issue, which is great. Just need to think a bit more about whether we would be suppressing any potential painful bugs by ignoring errors here.

File: languages/rust/oso/tests/test_polar_rust.rs [hunk 1]
```diff
@@ -110,6 +110,44 @@ fn test_load_function() {
...
+#[test]
+fn test_type_mismatch_fails_unification() {
+    common::setup();
+
+    #[derive(Eq, PartialEq, PolarClass, Clone)]
+    struct Foo {}
+    #[derive(Eq, PartialEq, PolarClass, Clone)]
+    struct Bar {}
+
+    impl Foo {}
```
[the remark this answers is not in the document]
Comment (AnIrishDuck) [reply languages/rust/oso/tests/test_polar_rust.rs:122]: I must've missed these when I was doing my final cleanup. They had methods I was using for testing at one point. Will remove.

File: languages/rust/oso/tests/test_polar_rust.rs [hunk 2]
```diff
@@ -110,6 +110,44 @@ fn test_load_function() {
...
+    impl Bar {}
+
+    let mut test = OsoTest::new();
+    test.oso
+        .register_class(
+            Foo::get_polar_class_builder()
+                .set_constructor(|| Foo {})
```
[the remark this answers is not in the document]
Comment (AnIrishDuck) [reply languages/rust/oso/tests/test_polar_rust.rs:129]: Yup, can do.

**Comment by AnIrishDuck** [comment]
I'm cool with lifting the check up to `Class::equals`. A preliminary `rg` indicates that's the only caller of `equality_check` currently, but I'm not sure if there's any reason that will remain true in the future.

I share your concern about suppressing errors and unintended consequences. One thing I found myself wondering was where on the strong / weak typing spectrum Polar lies as a language. I eventually decided the cheeky answer to "is Polar a strongly or weakly typed language?" is probably "yes". The top answer [here](https://stackoverflow.com/a/2785627) seems to agree.

I agree that multiple equality checks seems ideal in the longer term. The obvious argument for this is that the `Eq` and `PartialEq` traits in Rust support comparisons between different types.

**Comment by AnIrishDuck** [comment]
Looking at the other [now-failing spec](https://github.com/osohq/oso/blob/main/languages/rust/oso/tests/test_errors.rs#L59), I think an additional question is raised by [line 29](https://github.com/osohq/oso/blob/main/languages/rust/oso/tests/test_errors.rs#L29).

Specifically, should this logic be extended to Rust types that have no equality checks defined at all? Rephrased: should such types always fail to unify against every possible value? I'd lean towards "yes", but can think of arguments for "no".

My strongest argument for "yes" is that changing one of those values to `1` removes the error. That seems to indicate that at the very least, if the answer here is "no" the comparison of a built-in to a host type with no equality check should throw an error as well.

My thinking for "yes" stems from the assumption that host types should behave as similarly as possible to built-in Polar types. My additional reasoning  is that the `=` operator, unlike e.g. the `in` operator, is somewhat special in logic programming languages like Polar. It is strongly typed in the matches it will accept, but weakly typed in the comparisons it will allow.

It's been a while since I've worked with logic languages, so my reasoning could be faulty. I also realize we might be invoking wider design decisions that the team might not want to tackle now. I'm open to e.g. leaving the spec with commented asserts and backing out the other changes.

**Comment by samscott89** [comment]
Thanks for doing more thinking/investigating!

Polar is currently very much dynamically typed. Not sure how much the strong/weak spectrum makes sense.

You raise some really good points. Seems like there's a bunch of scenarios:

- unification in the head of rules. This _cannot_ be a hard error for comparing different types or it breaks the language
- equality checking `==` for types that have implemented eq. these could be a hard error
- equality checking `==` for types that haven't implemented eq. these could be  hard error

We could maybe make the behaviour of 1 differ from 2,3. I.e. if you use `==` then it _will_ be a runtime error if the types aren't comparable. But I worry about making unification and equality comparisons diverge like this.

On the other hand, we make them all fail/backtrack instead of error and risk hiding an implementation problem (you need to implement eq here!).

I think we should make `T vs U` be a fail always for now (In the future we can look for an equality). And T vs T be a hard error if the partialeq doesn't exist, since this only occurs in situations where someone really would want to do the comparison.

Which is roughly what you have, but it would be good to move that type check to somewhere more explicit.

**Comment by AnIrishDuck** [comment]
Got it. I'll make the type check explicit in `Class::equals` and make sure there are tests for the three enumerated cases.

**Comment by AnIrishDuck** [comment]
I believe I have all bases covered now. I've moved the type check and temporarily removed the `Foo` v `EqFoo` test in `test_errors.rs`. I'm pretty sure that support for multiple equality checks for different types will be needed before it can return.

While I can now imagine the mechanics of how such a feature could work, I think there are lots of other details that would require more thought. It definitely is out of scope for this PR.

File: languages/rust/oso/src/host/class.rs [hunk 3]
```diff
@@ -110,7 +110,11 @@ impl Class {
     }
 
     fn equals(&self, host: &Host, lhs: &Instance, rhs: &Instance) -> crate::Result<bool> {
-        (self.equality_check)(host, lhs, rhs)
+        if lhs.type_id() != rhs.type_id() {
```
Comment (samscott89) [inline languages/rust/oso/src/host/class.rs:113]:
```suggestion
        // equality checking is currently only supported for exactly matching types
        // TODO: support multiple dispatch for equality
        if lhs.type_id() != rhs.type_id() {
```

**Review by samscott89 — APPROVED** [review]
Thanks for helping to reason that one through!

**Pull request merged**